plans/TOOLING.md

```diff
@@ -0,0 +1,241 @@
+# THIS IS A WORK IN PROGRESS
+
+This document describes the tentative roadmap for AsyncAPI tooling in the next ~6 months and the strategies we'll follow to create the tools.
+
+## Roadmap
+We want to start developing tools as we make progress with version 2.0.0. To minimize the impact of future specification changes in the tools (_just those for v2 while it's in development_), we'll perform [all the breaking changes of the spec](https://github.com/asyncapi/asyncapi/issues?utf8=%E2%9C%93&q=label%3A%22v2+breaking+change%22+) first, and then we'll start with tooling **development**. Notice I've used the word _development_ because I want to make it clear that we can start other types of work beforehand, like planning, architecting, documenting the approach, etc. Actually, this document is already a pre-development work.
+
+### Milestones
+
+#### :arrow_right: Create a Parser/Validator
+The first tool we would have to implement is the AsyncAPI parser/validator. The parser/validator is a core piece of all the other tooling we'll build in the future. Therefore it's a top priority for us.
+
+##### Requirements:
+1. It MUST support YAML and JSON as input formats.
+2. It MUST support the following schema formats: OpenAPI schemas, JSON Schema Draft 04-07, Protobuf, and Avro.
```

Comment:
or am I missing some basic understanding here? for me validation of spec against schema is always spec as input document, schema as the schema that document must be validated against and that is it. So I see it in the way that you pass the document, from the document we know if it is following 1.2 or 2 and validate against https://github.com/asyncapi/asyncapi/blob/master/versions/1.2.0/schema.json for example. Like in this super simple library https://github.com/xeipuuv/gojsonschema

Reply:
Yes, but since we're introducing different schema/message formats like Avro and Protobuf, a JSON Schema validator is not enough. Check out this issue for context: https://github.com/asyncapi/asyncapi/issues/113.

Reply:
ok, so if I get it right asyncapi org would maintain asyncapi schema in different formats? deliver for example json schema and then also transform and officially support other formats? sorry for dumb questions but I want to make sure I fully get it

Reply:
Not the whole document, just the message payloads and headers. The following is an example using protobuf.

##### asyncapi.yaml
```yaml
asyncapi: '2.0.0'
info:
  title: 'My API'
  version: '1.0.0'

channels:
  user/created:
    subscribe:
      headers:
        application/x-protobuf:
          $ref: 'path/to/user-created.proto#Headers'
      payload:
        application/x-protobuf:
          $ref: 'path/to/user-created.proto#UserCreated'
```

##### user-created.proto

```proto
message Headers {
  string my-app-header = 1;
}

message UserCreated {
  int32 id = 1;
  string name = 2;
  string email = 3;
}
```

Notice the whole document remains the same. It's just the message (publish/subscribe) part that changes. The `asyncapi.yaml` example above would be equivalent to this one:

```yaml
asyncapi: '2.0.0'
info:
  title: 'My API'
  version: '1.0.0'

channels:
  user/created:
    subscribe:
      headers:
        application/schema+json;version=draft-07:
          type: object
          properties:
            'my-app-header':
              type: string
      payload:
        application/schema+json;version=draft-07:
          type: object
          properties:
            id:
              type: integer
            name:
              type: string
            email:
              type: string
```

Does it make sense?